### DISCUSSION TITLE
How can I see gRPC server streaming sessions?

### DISCUSSION THREAD

**Discussion by ncsibra** [AnWeber/vscode-httpyac#231] [Q&A] [answered]
Hi,

Where can I see the active gRPC server streaming connections?
They're working fine, I get the messages in the output view when choosing "httpyac - GRPC", but otherwise can't find them anywhere.
Not shown in the "Session" view and it's not even in the "History", only after closing the connection.
Checked with httpbook too, but didn't see any info about those.

It's should be shown somewhere, most likely in "Session" or the "active session" view in httpbook.
Currently I don't know how to terminate them (currently closing VSCode) and don't know how many of them are still active without getting a message.

**Comment by AnWeber** [comment]
Hi I agree with you. The behavior is still not quite ideal. For individual clients I already use the session store for storage and reuse (e.g. [AMQP](https://github.com/AnWeber/httpyac/blob/main/src/plugins/rabbitmq/amqpRequestClient.ts#L64)), but not for all. I think it's best to implement this generically.
Currently, however, a popup should actually be open that shows the running process abortable. Is this not working correctly?
<img width="298" alt="image" src="https://github.com/AnWeber/vscode-httpyac/assets/13434739/eb6dbee3-82ab-469f-a641-eb9478691005">

**Reply by ncsibra** [reply]
There's a popup when starting the server streaming request, which is gone after a few seconds, like most of the popups.
Maybe it's an issue only with gRPC?

**Reply by AnWeber** [reply]
The popup is currently the only way to cancel the long running streams. If it disappears, that's not really good. It is also not a problem explicitly with grpc. AMQP is the special one,  i use more actively. But it would make sense to treat all request clients (grpc, amqp, websocket,...) the same. I am supposed to love all my children equally even though I struggle with httpbook:-)

**Reply by ncsibra** [reply]
I tested again, with httpbook and without, but in both cases, the popup was gone.
Making the popup work would be a good start, but in the case of multiple streams, the popups might clutter the rest of the editor.
So using a generic solution for all of your children, putting these in the session store, like you already mentioned would be the best indeed. :)

**Reply by AnWeber** [reply]
Solution creating a session for all connections will be added with Anweber/httpyac#565.

**Reply by ncsibra** [reply]
Thanks for the fix, tested it, but it does not really work.

In text editor mode, when I start a server streaming it will be added to the sessions view and a pop-up appears with a cancel button (I didn't have this before, even though you said I should :) ).
When I press "X" in sessions or press the cancel button on the popup, they disappear but none of them disconnect from the server.
I don't get messages anymore in the output even if the server sends them, but get a message when the server closes the connection, e.g. times out, which proves that it was connected on the VSCode side too.
Btw, the popup and session stay in case the server closes the connection.
Only see a disconnect on the server side after closing the VSCode.

In httpbook, when I press the stop button on the cell, it does not do anything, no disconnect, progress bar and duration counter go on, pressing stop again does not help.
When I delete the connection from the session view the same happens, the cell still shows the progress bar/counter/stop button.
The other issue with httpbook is that I don't see a way to start multiple streaming sessions on the same endpoint, because the cell only shows the stop button.
In text editor mode I can press "connect" multiple times and it creates multiple sessions, don't know how to achieve the same in httpbook.

Tested our endpoint with Postman too, just to be sure that the issue is not on our side, but that works correctly, when I press the "Cancel" button it disconnects from the server and I can see that on the server side too.

**Reply by AnWeber** [reply]
> I don't get messages anymore in the output even if the server sends them, but get a message when the server closes the connection, e.g. times out, which proves that it was connected on the VSCode side too.

I overlooked that the client expects an explicit `close` call. This has caused a memory leak. I have added this now. And released a new version

> Btw, the popup and session stay in case the server closes the connection.

I listened to stream event [`end`](https://nodejs.org/api/stream.html#event-end). The correct approach, however, would have been [`close`](https://nodejs.org/api/stream.html#event-close). I have now also corrected this in the current release.

> In httpbook, when I press the stop button on the cell, it does not do anything, no disconnect, progress bar and duration counter go on, pressing stop again does not help.

Event `end` vs `close` error. Stream gets closed, but Promise is not resolved. Fixed with change to correct event.

> The other issue with httpbook is that I don't see a way to start multiple streaming sessions on the same endpoint, because the cell only shows the stop button.

This is not possible by means of the Run button of a cell. This is a limitation of the VSCode notebooks that I cannot prevent. It is possible to enable codelens display in code cells as well (setting `httpyac.useMethodInSendCodeLens`)

**Reply by ncsibra** [reply]
This fixed the UI part of the issue, so the stop button working now, and removing the session removes the popup too, on a server close both the session and popup disappear.
The only issue is that it still not actually closing the connection, don't see it on the server side and if I close the connection on the server after removing the session in VSCode I still get a "Call cancelled" message in the output, which can only happen if the connection is still active.

**Reply by AnWeber** [reply]
@ncsibra I need help with that. I can't readjust/check it and to me the behavior looks right. I am now correctly calling [client.close](https://github.com/AnWeber/httpyac/blob/main/src/plugins/grpc/grpcRequestClient.ts#L177) ([docs](https://grpc.github.io/grpc/node/grpc.Client.html#close__anchor)). The [connectivitystate](https://grpc.github.io/grpc/node/grpc.html#.connectivityState) before the close is 2 = Ready, after close 4 = shutdown. And that actually all looks correct and I do not know what more I need to do.. Could you provide me a server and the necessary steps how to reproduce the error?
I'll make one more attempt now without an example. There is a [cancel method on the stream](https://github.com/grpc/grpc-node/blob/779e970099b155f92b54a594c1b40fcef4378031/packages/grpc-js/src/call.ts#L114). Maybe I have to call this instead of the destroy I expect.

**Reply by ncsibra** [reply]
I will try to create a small sample server, but the cancel method sounds promising, so worth a try.

**Reply by ncsibra** [reply]
@AnWeber Here's a sample server: https://github.com/ncsibra/grpc-examples/tree/main/route_guide
It's written in Go, if you need a binary instead, let me know the platform, and I will create one.
I added a very simple sample `.http` file too.
After calling `ListFeatures` it will send a message every 100s, 100 messages in total, logging after every sending and will log when it finished or the client cancelled the call.

After some testing pretty sure that calling cancel will be the solution.
When testing in Postman I can see the cancelled log in the server, but when removing the session in httpyac it just logs the sendings until finished.

**Reply by AnWeber** [reply, marked as answer]
I have now set it up in a dev container. Thanks for the example. And yes calling the cancel method seems to solve the problem. Without it the server rattles through all 100 entries. Now when I use my dev branch it stops after cancel. 

```sh
2023/10/23 19:12:22 22. feature sent
2023/10/23 19:12:23 23. feature sent
2023/10/23 19:12:23 context canceled during sleep
2023/10/23 19:12:23 ListFeatures finished
```

Apparently it is not enough for grpc to terminate the stream using destroy/ end, but you have to explicitly use the API. But cool found I hope. I will release the state and then you can still test it please. And big thanks for your example. It was super helpful.

**Reply by ncsibra** [reply]
Looks like it's working correctly, thanks for the fix!
